AccountingSystem/AccountingSystem/CSVHandler.h

```diff
@@ -0,0 +1,176 @@
+#pragma once
+#include <string>
+#include <functional>
+#include <vector>
+#include <cstdio>
+#include <experimental/filesystem>
+#include "IDataHandler.h"
+
+template <class T>
+class CSVHandler : public IDataHandler<T>
+{
+public:
+	 CSVHandler(std::string pathToFile, std::vector<std::string> columnsName, std::function<std::string(T)> convertToInputFormat,
+	            std::function<T(std::string)> convertToOutputFormat);
+	 int Add(const T data) override;
+	 int Delete(const int id) override;
+	 int Update(const int id, const T value) override;
+	 T Get(const int id) override;
+     T GetLastRecord() override;
+     int TableSize() override;
+private:
+	void CreateFile(const std::string pathToFile);
+    void CreateTmpFile(const std::string pathToFile, std::vector<std::string> rows);
+    std::string searchRow(int id);
+	std::function<std::string(T)> _toInputFormat;
+	std::function<T(std::string)> _toOutputFormat;
+	std::vector<std::string> _columnsName;
+	std::string _pathToFile;
+};
+
+template<class T>
+CSVHandler<T>::CSVHandler(std::string pathToFile, std::vector<std::string> columnsName, std::function<std::string(T)> convertToInputFormat,
+    std::function<T(std::string)> convertToOutputFormat)
+    : _pathToFile(pathToFile), _columnsName(columnsName), _toInputFormat(convertToInputFormat),
+    _toOutputFormat(convertToOutputFormat) {}
+
+template<class T>
+inline int CSVHandler<T>::Add(const T data)
+{
+     if (!std::experimental::filesystem::exists(_pathToFile))
+     {
+          CreateFile(_pathToFile);
+     }
+     std::ofstream fout;
+     fout.open(_pathToFile, std::ios_base::app);
+     std::string result = _toInputFormat(data);
+     fout << result << "\n";
+     fout.close();
+     return 0;
+}
+
+template<class T>
+inline int CSVHandler<T>::Delete(const int id)
+{
+     if (!std::experimental::filesystem::exists(_pathToFile))
```

Comment:
good that you haven't put it in construct. The code looks more ugly this way but it's safer, since create file might throw. And throwing in the constructor is bad